File: sections/72_master_flow_convex_mixed/problems/c_selective_quota_profit/solution.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
const ll INF = (1LL << 62);

struct MCF {
    struct Edge { int to, rev, cap; ll cost; };
    vector<vector<Edge>> graph;
    explicit MCF(int n) : graph(n) {}

    pair<int, int> add_edge(int u, int v, int cap, ll cost) {
        int index = graph[u].size();
        graph[u].push_back({v, (int)graph[v].size(), cap, cost});
        graph[v].push_back({u, index, 0, -cost});
        return {u, index};
    }

    pair<int, ll> flow(int source, int sink, int limit) {
        int n = graph.size();
        vector<ll> potential(n, INF), distance(n);
        potential[source] = 0;
        for (int repeat = 0; repeat < n; ++repeat) {
            bool changed = false;
            for (int u = 0; u < n; ++u) {
                if (potential[u] == INF) continue;
                for (const Edge& edge : graph[u]) {
                    if (edge.cap
                        && potential[edge.to] > potential[u] + edge.cost) {
                        potential[edge.to] = potential[u] + edge.cost;
                        changed = true;
                    }
                }
            }
            if (!changed) break;
        }
        for (ll& value : potential)
            if (value == INF) value = 0;

        int sent = 0;
        ll total_cost = 0;
        vector<int> parent_v(n), parent_e(n);
        while (sent < limit) {
            fill(distance.begin(), distance.end(), INF);
            distance[source] = 0;
            priority_queue<pair<ll, int>,
                vector<pair<ll, int>>, greater<pair<ll, int>>> queue;
            queue.push({0, source});
            while (!queue.empty()) {
                auto [current, u] = queue.top();
                queue.pop();
                if (current != distance[u]) continue;
                for (int i = 0; i < (int)graph[u].size(); ++i) {
                    const Edge& edge = graph[u][i];
                    if (!edge.cap) continue;
                    ll next = current + edge.cost
                            + potential[u] - potential[edge.to];
                    if (next >= distance[edge.to]) continue;
                    distance[edge.to] = next;
                    parent_v[edge.to] = u;
                    parent_e[edge.to] = i;
                    queue.push({next, edge.to});
                }
            }
            if (distance[sink] == INF) break;
            for (int v = 0; v < n; ++v)
                if (distance[v] < INF) potential[v] += distance[v];
            int add = limit - sent;
            for (int v = sink; v != source; v = parent_v[v])
                add = min(add,
                    graph[parent_v[v]][parent_e[v]].cap);
            for (int v = sink; v != source; v = parent_v[v]) {
                Edge& edge = graph[parent_v[v]][parent_e[v]];
                total_cost += ll(add) * edge.cost;
                edge.cap -= add;
                graph[v][edge.rev].cap += add;
            }
            sent += add;
        }
        return {sent, total_cost};
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, projects, m, chosen;
    cin >> n >> projects >> m >> chosen;
    vector<int> low(projects), high(projects);
    int mandatory_count = 0;
    for (int j = 0; j < projects; ++j) {
        cin >> low[j] >> high[j];
        mandatory_count += low[j];
    }
    struct Choice { int worker, project, profit; };
    vector<Choice> choices(m);
    int maximum_profit = 0;
    for (Choice& choice : choices) {
        cin >> choice.worker >> choice.project >> choice.profit;
        --choice.worker;
        --choice.project;
        maximum_profit = max(maximum_profit, choice.profit);
    }

    int source = n + projects;
    int sink = source + 1;
    MCF network(sink + 1);
    for (int worker = 0; worker < n; ++worker)
        network.add_edge(source, worker, 1, 0);
    for (const Choice& choice : choices)
        network.add_edge(choice.worker, n + choice.project,
                         1, -choice.profit);

    ll big = ll(chosen + 1) * (maximum_profit + 1LL);
    vector<pair<int, int>> mandatory;
    for (int project = 0; project < projects; ++project) {
        mandatory.push_back(network.add_edge(
            n + project, sink, low[project], -big));
        network.add_edge(n + project, sink,
                         high[project] - low[project], 0);
    }

    auto [sent, cost] = network.flow(source, sink, chosen);
    bool possible = sent == chosen;
    for (auto [u, index] : mandatory)
        possible &= network.graph[u][index].cap == 0;
    if (!possible)
        cout << "IMPOSSIBLE\n";
    else
        cout << -cost - big * mandatory_count << '\n';
}

File: sections/72_master_flow_convex_mixed/problems/c_selective_quota_profit/solution.py
import heapq
import sys


class MCF:
    def __init__(self, n):
        self.graph = [[] for _ in range(n)]

    def add_edge(self, u, v, capacity, cost):
        reference = (u, len(self.graph[u]))
        self.graph[u].append([v, capacity, cost, len(self.graph[v])])
        self.graph[v].append([u, 0, -cost, len(self.graph[u]) - 1])
        return reference

    def flow(self, source, sink, limit):
        n = len(self.graph)
        infinity = 10**40
        potential = [infinity] * n
        potential[source] = 0
        for _ in range(n):
            changed = False
            for u in range(n):
                if potential[u] == infinity:
                    continue
                for v, capacity, cost, _ in self.graph[u]:
                    if capacity and potential[v] > potential[u] + cost:
                        potential[v] = potential[u] + cost
                        changed = True
            if not changed:
                break
        potential = [0 if value == infinity else value for value in potential]

        sent = total_cost = 0
        while sent < limit:
            distance = [infinity] * n
            parent = [None] * n
            distance[source] = 0
            queue = [(0, source)]
            while queue:
                current, u = heapq.heappop(queue)
                if current != distance[u]:
                    continue
                for index, edge in enumerate(self.graph[u]):
                    v, capacity, cost, _ = edge
                    if not capacity:
                        continue
                    candidate = current + cost + potential[u] - potential[v]
                    if candidate < distance[v]:
                        distance[v] = candidate
                        parent[v] = (u, index)
                        heapq.heappush(queue, (candidate, v))
            if distance[sink] == infinity:
                break
            for v in range(n):
                if distance[v] < infinity:
                    potential[v] += distance[v]
            add = limit - sent
            vertex = sink
            while vertex != source:
                u, index = parent[vertex]
                add = min(add, self.graph[u][index][1])
                vertex = u
            vertex = sink
            while vertex != source:
                u, index = parent[vertex]
                edge = self.graph[u][index]
                total_cost += add * edge[2]
                edge[1] -= add
                self.graph[vertex][edge[3]][1] += add
                vertex = u
            sent += add
        return sent, total_cost


def main():
    data = list(map(int, sys.stdin.buffer.read().split()))
    n, projects, m, chosen = data[:4]
    at = 4
    bounds = []
    for _ in range(projects):
        bounds.append(tuple(data[at:at + 2]))
        at += 2
    choices = []
    maximum_profit = 0
    for _ in range(m):
        worker, project, profit = data[at:at + 3]
        at += 3
        choices.append((worker - 1, project - 1, profit))
        maximum_profit = max(maximum_profit, profit)

    source = n + projects
    sink = source + 1
    network = MCF(sink + 1)
    for worker in range(n):
        network.add_edge(source, worker, 1, 0)
    for worker, project, profit in choices:
        network.add_edge(worker, n + project, 1, -profit)

    big = (chosen + 1) * (maximum_profit + 1)
    mandatory = []
    for project, (low, high) in enumerate(bounds):
        mandatory.append(
            network.add_edge(n + project, sink, low, -big)
        )
        network.add_edge(n + project, sink, high - low, 0)

    sent, cost = network.flow(source, sink, chosen)
    possible = sent == chosen and all(
        network.graph[u][index][1] == 0 for u, index in mandatory
    )
    if possible:
        print(-cost - big * sum(low for low, _ in bounds))
    else:
        print("IMPOSSIBLE")


if __name__ == '__main__':
    main()
